### PR TITLE
pydarshan: new APMPI stats in a html report generation

### PR DESCRIPTION
**Pull request by sudheer9**
changes for summarizing APMPI from Darshan into new job summary report

### CODE REVIEW

**Comment by shanedsnyder** [comment]
@tylerjereddy FYI, this is a cleaned up PR with @sudheer9 changes to darshan-job-summary tool to account for some APMPI data.

I know we're still sorting out what we'll ultimately use for the report generation, but this is just an example of the types of information we're already interested in summarizing for the APMPI module. I don't know that any serious review here is needed on your end, but I'll leave the PR open for a bit in case you have any comments before I merge in underneath you.

**Comment by tylerjereddy** [comment]
Ok, if you have a sample file with the appropriate tracing data (or is APMPI tracing available in the test suite already?) maybe Nik could take a look. I'm trying to avoid being the only person doing detailed code review--since stuff isn't going into `main` anyway, it is probably a good opportunity for folks to get used to reviewing each other (and responding to/addressing review comments) with fairly low stakes initially.

**Comment by sudheer9** [comment]
The sample report the gets generated with APMPI stats is here: https://github.com/sudheer9/misc/blob/master/report.html.

**Comment by tylerjereddy** [comment]
I was looking for a darshan log file with APMPI tracing data in it; in any case, I'll generate one myself so I can at least run the code.

**Comment by sudheer9** [comment]
here it is: https://github.com/sudheer9/misc/blob/master/526820.2.64p.log

**Review by tylerjereddy — COMMENTED** [review]
When I use this feature branch locally:

`python -m darshan.cli summary --output=report.pdf --input=526820.2.64p.log `

I get:

`Segmentation fault (core dumped)`

Before that, I was getting:

```
AttributeError: module 'darshan.backend.cffi_backend' has no attribute 'log_get_apmpi_record'
```

So, I rebuilt `darshan-util` via `./configure --prefix=/home/tyler/darshan_install/lib --enable-autoperf-apmpi`

I'll check the target feature branch to see if it behaves any better in my hands with the same test log file.

**Comment by carns** [comment]
> here it is: https://github.com/sudheer9/misc/blob/master/526820.2.64p.log

Hi @sudheer9 , Ok with you if we put a copy of this log in the https://github.com/darshan-hpc/darshan-logs repository?

File: darshan-util/pydarshan/darshan/cli/summary.py [hunk 1]
```diff
@@ -272,6 +260,12 @@ def apmpi_process(apmpi_dict):
     pyplot.savefig(buf, format='png', bbox_inches='tight')
     buf.seek(0)
     encoded.append(base64.b64encode(buf.read()))
+    df_max_rank = df_apmpi.loc[df_apmpi["Rank"] == max_rank]
+    df_min_rank = df_apmpi.loc[df_apmpi["Rank"] == min_rank]
+    df_mean_rank = df_apmpi.loc[df_apmpi["Rank"] == mean_rank]
+    encoded.append(df_max_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    encoded.append(df_min_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    encoded.append(df_mean_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
```
[the remark this answers is not in the document]
Comment (sudheer9) [reply darshan-util/pydarshan/darshan/cli/summary.py:268]: Agreed.

**Comment by sudheer9** [comment]
Hi @carns sure, I can place it there.

**Comment by sudheer9** [comment]
> When I use this feature branch locally:
> 
> `python -m darshan.cli summary --output=report.pdf --input=526820.2.64p.log `
> 
> I get:
> 
> `Segmentation fault (core dumped)`
> 
> Before that, I was getting:
> 
> ```
> AttributeError: module 'darshan.backend.cffi_backend' has no attribute 'log_get_apmpi_record'
> ```
> 
> So, I rebuilt `darshan-util` via `./configure --prefix=/home/tyler/darshan_install/lib --enable-autoperf-apmpi`
> 
> I'll check the target feature branch to see if it behaves any better in my hands with the same test log file.

@tylerjereddy can you try the updated code again, it at least tells me if it had made any progress or pointers on where it is seg faulting.

**Comment by tylerjereddy** [comment]
@sudheer9 It still segfaults, and on other branches when using that test file as well. It might be helpful if other folks want to confirm if they can/can't process your sample file in a simple way like this two-line reproducer:

```python
import darshan

report = darshan.DarshanReport('526820.2.64p.log', read_all=True)
```

maybe @nawtrey can check--perhaps good exercise to make sure you can build `darshan-util` with the appropriate APMPI support as well (maybe that's part of the problem here, not sure, but segfault shouldn't happen even if there is a developer install error ideally).

**Comment by tylerjereddy** [comment]
for diagnostic purposes `read_all=False` avoids the segfault, perhaps not too surprising

**Comment by sudheer9** [comment]
`./configure --enable-autoperf-apmpi --enable-shared --enable-pydarshan CC=gcc` - this is how I build `darshan-util` and do not see the segfault. yeah, it would be good to test.

**Comment by tylerjereddy** [comment]
Using only`./configure --prefix=/home/tyler/... --enable-autoperf-apmpi` the segfault is definitely there--probably there are missing guards on Python/C API calls and something that doesn't exist is getting called instead of raising an error.

**Comment by sudheer9** [comment]
ok. So, you are not using `--enable-pydarshan` right.
Also, does using `darshan-parser` on the log with APMPI data works fine for you?

**Comment by tylerjereddy** [comment]
I suspect it won't segfault if built correctly, but we might as well have an error message telling us what to do in that case instead of a segfault. I'll try building `darshan-util` with a few other incantations.

What are all the other changes you've pushed into this PR now? The diff is getting pretty big, and I see some duplicated imports added in?

**Comment by sudheer9** [comment]
ok.

I could have waited before pushing those new changes, it is still WIP, basically adding more analysis plots for the APMPI log data. I pushed those have now thinking this version may not lead to Seg fault, but the problem could be elsewhere. I will cleanup the duplicated imports; My progress was help up because of the huge time being taken for saving a `plotnine` figure and I am figuring out ways to fix that.

**Comment by tylerjereddy** [comment]
Ok, I can avoid the segfault if I build `darshan-util` with all the special flags. I still get an error, but I'll merge this to the main PR and add more comments there...

**Pull request merged**